bluetooth: tester: btp_gap: fix building without extended adv

kedMertens opened the pull request:
btp_gap implementation references ext_adv, which is guarded by if defined macro. Disabling CONFIG_BT_EXT_ADV will fail compilation. Fixed it by moving declaration out of the guards, it will be dropped by linker if not used and marking as maybe_unused, to prevent warning if kconfig is not set. 
set_extended_advertising moved under the guard to avoid unused warning.

jhedberg commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -560,6 +561,28 @@ int tester_gap_stop_ext_adv(void)
+
+	LOG_DBG("ext adv settings: %u", cp->settings);
+
+	if (cp->settings != 0) {
+		atomic_set_bit(&current_settings,
+			       BTP_GAP_SETTINGS_EXTENDED_ADVERTISING);
+	} else {
+		atomic_clear_bit(&current_settings,
+				 BTP_GAP_SETTINGS_EXTENDED_ADVERTISING);
+	}
```
I know this is just copying existing code and thereby potentially out of scope to this PR, however such code patterns can generally be replaced by a single line:
```c
atomic_set_bit_to(&current_settings, BTP_GAP_SETTINGS_EXTENDED_ADVERTISING,
		  cp->settings != 0);
```

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -521,8 +521,9 @@ static struct bt_data ad[10] = {
 };
 static struct bt_data sd[10];
 
+__maybe_unused static struct bt_le_ext_adv *ext_adv;
```
Looking at the file, it is only this part that uses `IS_ENABLED`:
```c
	if (IS_ENABLED(CONFIG_BT_EXT_ADV) && atomic_test_bit(&current_settings,
	    BTP_GAP_SETTINGS_EXTENDED_ADVERTISING)) {
		param->options |= BT_LE_ADV_OPT_EXT_ADV;
		if (ext_adv != NULL) {
			err = bt_le_ext_adv_stop(ext_adv);
			if (err != 0) {
				return err;
			}

			err = bt_le_ext_adv_delete(ext_adv);
			if (err != 0) {
				return err;
			}

			ext_adv = NULL;
		}

		err = bt_le_ext_adv_create(param, NULL, &ext_adv);
		if (err != 0) {
			return BTP_STATUS_FAILED;
		}

		err = bt_le_ext_adv_set_data(ext_adv, ad, ad_len, sd_len ? sd : NULL, sd_len);
	}
```
from `tester_gap_create_adv_instance`; everywhere else it is guarded by `#if defined(CONFIG_BT_EXT_ADV)`. 

To me that indicates that the above part should be guarded by `#if defined(CONFIG_BT_EXT_ADV)` instead of `IS_ENABLED`, or that none of them should guard the access of it by `#if defined(CONFIG_BT_EXT_ADV)` and instead use `IS_ENABLED`. 

I don't think moving the declaration out and marking it as `__maybe_unused` is correct, as it is definitely not used if `CONFIG_BT_EXT_ADV=n`, unless I missed some code that may access it when `CONFIG_BT_EXT_ADV=y`.

kedMertens replied:
Good point, you are right everywhere else it is guarded, if moving this block under guards we would loose a benefit of checking for compilation errors, even-though it is the only place in the file. 
Is a desired way would to make use of IS_ENABLED in rest of the file or is it risky?

kedMertens replied:
Refactored with IS_ENABLED, I've run several GAP pts test cases, to make sure that nothing is broken by the change.

kedMertens committed:
bluetooth: tester: btp: refactors OOB legacy selection

Takes in use IS_ENABLED as preferred way of guarding Kconfig option for
forcing OOB legacy, CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY. It enables
compiling guarded code get warning/errors and not link it if not used. If
def macro is left in btp_handler structure to reduce code size when the
config is not present.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

kedMertens committed:
bluetooth: tester: btp: refactors privacy selection

Takes in use IS_ENABLED as preferred way of guarding Kconfig option for
enabling privacy, CONFIG_BT_PRIVACY. It enables compiling guarded code get
warning/errors and not link it if not used.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

kedMertens committed:
bluetooth: tester: btp: fixes ext adv selection

Disabling extended advertising, CONFIG_BT_EXT_ADV, fails to compile cause
references not guarded ext_adv variable when tester_gap_create_adv_instance
is called. Fixes by taking in use IS_ENABLED as preferred way of guarding
Kconfig option. It enables compiling guarded code to get warning/errors and
not to link if not used.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

kedMertens committed:
bluetooth: tester: btp: refactors LE secure pairing selection

Takes in use IS_ENABLED as preferred way of guarding Kconfig option for
enabling legacy LE secure pairing, CONFIG_BT_SMP_SC_PAIR_ONLY. It enables
compiling guarded code get warning/errors and not link it if not used.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

kedMertens committed:
bluetooth: tester: btp: refactors periodic adv selection

Takes in use IS_ENABLED as preferred way of guarding Kconfig option for
enabling periodic advertising api, CONFIG_BT_PER_ADV. It enables compiling
guarded code get warning/errors and not link it if not used.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

sjanc commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1527,6 +1561,9 @@ int tester_gap_padv_start(void)
 static uint8_t padv_start(const void *cmd, uint16_t cmd_len,
 			  void *rsp, uint16_t *rsp_len)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
we could just not register those at all instead of fake response from registered handler

kedMertens replied:
They are guarded in handlers and will not be registered. You are right it should be here in that case as it can't be called from anywhere else, I'll remove it.

sjanc reviewed (commented):
I'm not sure which path is better but we should at least be consistent... 

I don't think IS_ENABLED() is right tool for disabling BTP handlers, we should just ifdef whole code and don't register handler in first place   (but I guess this depends on person being asked :P)

kedMertens commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1600,6 +1646,9 @@ int tester_gap_padv_set_data(struct bt_data *per_ad, uint8_t ad_len)
 static uint8_t padv_set_data(const void *cmd, uint16_t cmd_len,
 			     void *rsp, uint16_t *rsp_len)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
+		return BTP_STATUS_UNKNOWN_CMD;
+	}
```
@sjanc, are you referring to once that are marked as TODO?

kedMertens commented:
I've addressed suggestion, except unregistered btp handlers
It is tricky to guard with defined macro them in handler and not get warning -Wunused-function, using __maybeunused not suitable as will hide if it is really unused and not intended to be so. 
Suggestion was to revert and guard them too.
Other option would be to remove guards from handlers completely and then return based on configs from functions is it is enabled or disabled, is it a viable option or better to revert @sjanc ?

kedMertens commented:
- [ ] Merge fixup commits prior to merging

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -45,38 +45,28 @@ static uint8_t oob_legacy_tk[16] = { 0 };
 			  struct bt_gatt_read_params *params, const void *data,
 			  uint16_t length)
 {
+	if (IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		return -ENOSYS;
+	}
```
This doesn't make any sense :) It's a callback that returns a `uint8_t`

Should be `BT_GATT_ITER_STOP` if anything, but why is this being guarded at all? The callback does not seem related to privacy.

kedMertens replied:
I've checked git history, it was originally guarded to satisfy GAP/CONN/DCON/BV-05-C, it exercises privacy mode behavior, in d80e480f2005 it describes that:
> This allows to do directed advertising with peer address set to RPA.
> To do this according to specification IUT must first read Central
> Address Resolution characteristic to check if peer supports it.
I guess it is possible to return BT_GATT_ITER_STOP, but how one would see that there was an error?

Thalley replied:
> I've checked git history, it was originally guarded to satisfy GAP/CONN/DCON/BV-05-C, it exercises privacy mode behavior, in [d80e480](https://github.com/zephyrproject-rtos/zephyr/commit/d80e480f20050171d199fea79a1959bdbb75932d) it describes that:
> 
> > This allows to do directed advertising with peer address set to RPA.
> > To do this according to specification IUT must first read Central
> > Address Resolution characteristic to check if peer supports it.
> 
> I guess it is possible to return BT_GATT_ITER_STOP, but how one would see that there was an error?

You already guard the call that uses this callback, so if you want any check here, we might just add an assert. We could also just remove the check in the callback, as it should never happen

Thalley replied:
@kedMertens This still need to be changed :)

Either remove the check (my preference), add an ASSERT or return `BT_GATT_ITER_STOP` with a log statement

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -308,19 +305,19 @@ static uint8_t controller_info(const void *cmd, uint16_t cmd_len,
-#if !defined(CONFIG_BT_PRIVACY)
-	if (oob_local.addr.type == BT_ADDR_LE_RANDOM) {
-		atomic_set_bit(&current_settings, BTP_GAP_SETTINGS_STATIC_ADDRESS);
+	if (!IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		if (oob_local.addr.type == BT_ADDR_LE_RANDOM) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PRIVACY) && oob_local.addr.type == BT_ADDR_LE_RANDOM) {
```

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1705,14 +1774,17 @@ static uint8_t padv_sync_transfer_start(const void *cmd, uint16_t cmd_len,
 static uint8_t padv_sync_transfer_recv(const void *cmd, uint16_t cmd_len,
 				       void *rsp, uint16_t *rsp_len)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PER_ADV_SYNC_TRANSFER_RECEIVER)) {
```

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1683,6 +1744,10 @@ static uint8_t padv_create_sync(const void *cmd, uint16_t cmd_len,
 static uint8_t padv_sync_transfer_set_info(const void *cmd, uint16_t cmd_len,
 					   void *rsp, uint16_t *rsp_len)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PER_ADV_SYNC_TRANSFER_SENDER)) {
```

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1657,6 +1714,10 @@ int tester_gap_padv_stop_sync(void)
 static uint8_t padv_create_sync(const void *cmd, uint16_t cmd_len,
 				void *rsp, uint16_t *rsp_len)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PER_ADV_SYNC)) {
```

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1640,6 +1693,10 @@ int tester_gap_padv_create_sync(struct bt_le_per_adv_sync_param *create_params)
 
 int tester_gap_padv_stop_sync(void)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PER_ADV_SYNC)) {
```

kedMertens replied:
Would it make sense to add proposed options to handlers registration and btp_gap header too?

Thalley replied:
It's a bit curious, because e.g. `tester_gap_padv_stop_sync` is only called in a BAP Broadcast Sink context, which requires `CONFIG_BT_PER_ADV_SYNC` already, so it doesn't really make sense to guard the function call in those contexts at least.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1623,6 +1672,10 @@ static uint8_t padv_set_data(const void *cmd, uint16_t cmd_len,
 
 int tester_gap_padv_create_sync(struct bt_le_per_adv_sync_param *create_params)
 {
+	if (!IS_ENABLED(CONFIG_BT_PER_ADV)) {
```
```suggestion
	if (!IS_ENABLED(CONFIG_BT_PER_ADV_SYNC)) {
```

Thalley reviewed (commented):
I think the changes are looking good now. We should, however, add an entry in https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/tester/testcase.yaml with `CONFIG_BT_EXT_ADV=n` to verify this in CI

kedMertens committed:
bluetooth: tester: btp: adds testcase tester legacy adv

Adds testcase to build tester with forced legacy advertisements,
CONFIG_BT_EXT_ADV=n.

Signed-off-by: Evgenii Kosenko <Evgenii.Kosenko@silabs.com>

Thalley reviewed (changes requested):
I think it's nearly OK as is. Would like to get input from the maintainer @sjanc on whether the approach is what they want or not
